### PR TITLE
File based dynamic config client

### PR DESCRIPTION
**Pull request by yycptt**
This implements https://github.com/uber/cadence/issues/1587

File based dynamic config client will read config from a file and store it in memory. Periodically, it will check the config file to see if it has been updated. If so, read and parse the file again.

This PR only contains the client implementation and unit tests. Right now, the file based client is not used by the cadence service. I can do that in a separate PR if needed.

### CODE REVIEW

File: common/service/dynamicconfig/fileBasedClient.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	client := &fileBasedClient{
+		config: config,
+		doneCh: doneCh,
+		logger: logger,
+	}
+	if err := client.update(); err != nil {
+		return nil, err
+	}
+	go func() {
+		ticker := time.NewTicker(client.config.PollInterval)
```
Comment (vancexu) [inline common/service/dynamicconfig/fileBasedClient.go]: PollInterval may need a default 5s to be defensive.

File: common/service/dynamicconfig/fileBasedClient.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		for {
+			select {
+			case <-ticker.C:
+				err := client.update()
+				if err != nil {
+					client.logger.Error("Failed to update dynamic config", tag.Timestamp(time.Now()))
+				}
+				client.logger.Info("Updated dynamic config", tag.Timestamp(time.Now()))
```
Comment (vancexu) [inline common/service/dynamicconfig/fileBasedClient.go]: would be better to put in else

Comment (yycptt) [reply common/service/dynamicconfig/fileBasedClient.go]: Updated.

File: common/service/dynamicconfig/fileBasedClient.go (hunk 1 above)
Comment (yycptt) [reply common/service/dynamicconfig/fileBasedClient.go]: Added a validate function on start to check both poll interval and whether the config file exists.

**Pull request merged**